optimum/neuron/generation/utils.py

```diff
@@ -14,8 +14,1262 @@
+    decoder_hidden_states: Optional[Tuple[Tuple[torch.FloatTensor]]] = None
+
+
+GreedySearchOutput = Union[GreedySearchEncoderDecoderOutput, GreedySearchDecoderOnlyOutput]
+GenerateOutput = Union[GreedySearchOutput]
 
 
 class NeuronGenerationMixin(GenerationMixin):
-    pass
+    """
+    A class containing all functions for auto-regressive text generation on Trn1, to be used as a mixin in [`PreTrainedModel`].
+
+    The class exposes [`~generation.GenerationMixin.generate`], which can be used for:
+        - *greedy decoding* by calling [`~generation.GenerationMixin.greedy_search`] if `num_beams=1` and
+          `do_sample=False`
+        - *contrastive search* by calling [`~generation.GenerationMixin.contrastive_search`] if `penalty_alpha>0` and
+          `top_k>1`
+        - *multinomial sampling* by calling [`~generation.GenerationMixin.sample`] if `num_beams=1` and
+          `do_sample=True`
+        - *beam-search decoding* by calling [`~generation.GenerationMixin.beam_search`] if `num_beams>1` and
+          `do_sample=False`
+        - *beam-search multinomial sampling* by calling [`~generation.GenerationMixin.beam_sample`] if `num_beams>1`
+          and `do_sample=True`
+        - *diverse beam-search decoding* by calling [`~generation.GenerationMixin.group_beam_search`], if `num_beams>1`
+          and `num_beam_groups>1`
+        - *constrained beam-search decoding* by calling [`~generation.GenerationMixin.constrained_beam_search`], if
+          `constraints!=None` or `force_words_ids!=None`
+
+    You do not need to call any of the above methods directly. Pass custom parameter values to 'generate' instead. To
+    learn more about decoding strategies refer to the [text generation strategies guide](../generation_strategies).
+    """
+
+    def _update_model_kwargs_for_xla_generation(
+            self,
+            outputs: ModelOutput,
+            model_kwargs: Dict[str, Any],
+            batch_size: int,
+            is_encoder_decoder: bool = False,
+            standardize_cache_format: bool = False,
+            max_length: int = None,
+            seq_length: int = None,
+            use_cache=True,
+    ) -> Dict[str, Any]:
+        def _initialize_attention(model_kwargs, num_padding_values, is_encoder_decoder):
+            """initializes the appropriate attention mask -- encoder-decoder models use `decoder_attention_mask`"""
+            if is_encoder_decoder:
+                # One 1 for decoder_start_token_id, 0s for the currently-unfilled locations in the past_key_values tensor,
+                # 1s for the actual input_ids
+                decoder_attention_mask = torch.cat(
+                    [
+                        torch.zeros((batch_size, num_padding_values), dtype=torch.int32),
+                        torch.ones((batch_size, 2), dtype=torch.int32),
+                    ],
+                    axis=1,
+                ).to(outputs.logits.device)
```

Comment:
So we want the `decoder_attention_mask` to look something like `[[0, 0, 0, .., 1, 1]]`? (batch size = 1 in this example)
Shouldn't the ones come first?

Reply:
Alright